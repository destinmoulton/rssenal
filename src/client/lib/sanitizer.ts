import {
    SANITIZER_ALLOWED_CONTENT_TAGS,
    SANITIZER_ALLOWED_TAG_ATTRIBUTES
} from "../constants";

interface ISanitizeHTML {
    (html: string, options: any): string;
}
const sanitizeHTML: ISanitizeHTML = require("sanitize-html-react");

export const sanitizeEntryContent = (
    dirtyContent: string,
    shouldShowImages: boolean
): string => {
    let allowedTags = SANITIZER_ALLOWED_CONTENT_TAGS.slice();
    if (shouldShowImages) {
        allowedTags.push("img");
    }

    let transformTags = {
        a: (tagName: string, attribs: any) => {
            if (attribs.href && attribs.href[0] !== "/") {
                return {
                    tagName,
                    attribs: {
                        href: attribs.href,
                        target: "_blank"
                    }
                };
            } else {
                return {
                    tagName: "span"
                };
            }
        },
        b: "strong",
        h1: "strong",
        h2: "strong",
        h3: "strong",
        h4: "strong",
        h5: "strong",
        i: "em",
        img: (tagName: string, attribs: any) => {
            if (attribs.src[0] === "/") {
                // Don't display images with a relative path (as they don't exist on this server)
                return {
                    tagName: "span",
                    attribs: {}
                };
            } else {
                return {
                    tagName,
                    attribs: {
                        src: attribs.src,
                        class: "rss-entry-content-body-img"
                    }
                };
            }
        }
    };

    const tagsFilterIfEmpty = ["span", "p", "a", "em"];
    let exclusiveFilter = (frame: any) => {
        return tagsFilterIfEmpty.includes(frame.tag) && !frame.text.trim();
    };

    let sanitized = sanitizeHTML(dirtyContent, {
        allowedTags,
        SANITIZER_ALLOWED_TAG_ATTRIBUTES,
        exclusiveFilter,
        transformTags
    });

    // Finally, replace any stray <br>'s at the beginning
    return sanitized.replace(/^(<br \/>)*/, "");
};

export const sanitizeEntryTitle = (dirtyTitle: string) => {
    let allowedTags: string[] = [];
    let allowedAttributes: string[] = [];
    return sanitizeHTML(dirtyTitle, {
        allowedTags,
        allowedAttributes
    });
};
