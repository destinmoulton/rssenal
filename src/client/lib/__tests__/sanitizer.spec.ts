import { sanitizeEntryContent, sanitizeEntryTitle } from "../sanitizer";

import {
    INITIAL_HTML,
    EXPECTED_HTML_WITH_IMG,
    EXPECTED_HTML_WITHOUT_IMG
} from "../../../../test/data/sanitizer.data";

describe("lib/sanitizer.ts", () => {
    describe("sanitizeEntryContent() cleans up HTML tags", () => {
        it("transforms tags, with images", () => {
            const transformed = sanitizeEntryContent(INITIAL_HTML, true);
            expect(transformed).toBe(EXPECTED_HTML_WITH_IMG);
            expect(transformed).toMatchSnapshot();
        });
        it("transforms tags, stripping images", () => {
            const transformed = sanitizeEntryContent(INITIAL_HTML, false);
            expect(transformed).toBe(EXPECTED_HTML_WITHOUT_IMG);
            expect(transformed).toMatchSnapshot();
        });
    });

    describe("sanitizeEntryTitle()", () => {
        it("removes all tags", () => {
            const ORIG =
                "<span>Test</span><h1>one</h1><h2>two</h2><h3>three</h3><h4>four</h4><h5>five</h5><img src='bling'/><a href='test.com'>link</a>";
            const EXPECTED = "Testonetwothreefourfivelink";

            const transformed = sanitizeEntryTitle(ORIG);
            expect(transformed).toBe(EXPECTED);
        });
    });
});
