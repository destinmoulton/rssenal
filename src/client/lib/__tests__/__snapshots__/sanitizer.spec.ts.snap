// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/sanitizer.ts sanitizeEntryContent() cleans up HTML tags transforms tags, stripping images 1`] = `
"
<span>Gizoogle</span>
<a href=\\"http://www.gizoogle.com\\" target=\\"_blank\\">Gizoogle</a>


<strong>Test Bold</strong>
<strong>Test Bold</strong>

<br />

<blockquote>Test quotation</blockquote>

<pre><code>$(\\"test\\").html();</code></pre>

<strong>Test h1</strong>
<strong>Test h2</strong>
<strong>Test h3</strong>
<strong>Test h4</strong>
<strong>Test h5</strong>

<em>Test i</em>
<em>Test em</em>

<span></span>


<ol>
    <li>List Item</li>
</ol>
<ul>
    <li>List Item</li>
</ul>

<p>Test content</p>

<span>Test span</span>

<table>
    <thead>
        <tr>
            <th>Header Col 1</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>Body Col 1</td>
        </tr>
    </tbody>
</table>
"
`;

exports[`lib/sanitizer.ts sanitizeEntryContent() cleans up HTML tags transforms tags, with images 1`] = `
"
<span>Gizoogle</span>
<a href=\\"http://www.gizoogle.com\\" target=\\"_blank\\">Gizoogle</a>


<strong>Test Bold</strong>
<strong>Test Bold</strong>

<br />

<blockquote>Test quotation</blockquote>

<pre><code>$(\\"test\\").html();</code></pre>

<strong>Test h1</strong>
<strong>Test h2</strong>
<strong>Test h3</strong>
<strong>Test h4</strong>
<strong>Test h5</strong>

<em>Test i</em>
<em>Test em</em>

<span></span>
<img src=\\"http://test.com/test_image.png\\" />

<ol>
    <li>List Item</li>
</ol>
<ul>
    <li>List Item</li>
</ul>

<p>Test content</p>

<span>Test span</span>

<table>
    <thead>
        <tr>
            <th>Header Col 1</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>Body Col 1</td>
        </tr>
    </tbody>
</table>
"
`;
